pkg/addons/registry/migrate/pod.go

```diff
@@ -0,0 +1,371 @@
+package migrate
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"log/slog"
+	"regexp"
+	"time"
+
+	ecv1beta1 "github.com/replicatedhq/embedded-cluster/kinds/apis/v1beta1"
+	"github.com/replicatedhq/embedded-cluster/pkg/addons/seaweedfs"
+	"github.com/replicatedhq/embedded-cluster/pkg/kubeutils"
+	"github.com/replicatedhq/embedded-cluster/pkg/runtimeconfig"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	StatusConditionType = "RegistryMigrationStatus"
+
+	dataMigrationPodName = "registry-data-migration"
+	serviceAccountName   = "registry-data-migration-serviceaccount"
+
+	// seaweedfsS3SecretName is the name of the secret containing the s3 credentials.
+	// This secret name is defined in the chart in the release metadata.
+	seaweedfsS3SecretName = "seaweedfs-s3-rw"
+)
+
+// RunDataMigration should be called when transitioning from non-HA to HA airgapped installations.
+// This function creates a pod that will scale down the registry deployment then upload the data to s3
+// before finally creating a 'migration is complete' secret in the registry namespace. If this secret
+// is present, the function will return without reattempting the migration.
+func RunDataMigration(ctx context.Context, cli client.Client, kclient kubernetes.Interface, in *ecv1beta1.Installation, image string) (<-chan string, <-chan error, error) {
+	progressCh := make(chan string)
+	errCh := make(chan error, 1)
+
+	// TODO: this should be an argument
+	clientset, err := kubeutils.GetClientset()
+	if err != nil {
+		return nil, nil, fmt.Errorf("get kubernetes clientset: %w", err)
+	}
+
+	// TODO: should we check seaweedfs health?
+
+	logrus.Debug("Ensuring migration service account")
+	err = ensureMigrationServiceAccount(ctx, cli)
+	if err != nil {
+		return nil, nil, fmt.Errorf("ensure service account: %w", err)
+	}
+
+	logrus.Debug("Ensuring s3 access credentials secret")
+	err = ensureS3Secret(ctx, cli)
```

Comment:
this gets created by the seaweedfs chart, there's no need to do it here. it was needed previously when we used k0s extensions, but not anymore.

Reply:
actually, nevermind, this is in the registry namespace, not seaweedfs namespace. ignore

Reply:
hmm, it's not used in this migration though, and the registry chart creates it on upgrade. any reason to create it here?

Reply:
It is used here https://github.com/replicatedhq/embedded-cluster/pull/2027/files#diff-acb8fb3dc0e105f16b5eff96df7f6b3a3c3444c9bb9145965aa490608a06fed5R250